File: Semester 4/Programming and Projection Enironments/mpp-proiect-java/Proiect_MPP/AirportNetworking/src/main/java/root/rpcprotocol/ClientRpcReflectionWorker.java
package root.rpcprotocol;

import root.model.Airport;
import root.model.DTOFlight;
import root.model.Flight;
import root.model.Ticket;
import root.model.DTODestinationDate;
import root.model.people.Employee;
import root.services.IObserver;
import root.services.IService;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.util.Collection;

public class ClientRpcReflectionWorker implements Runnable, IObserver {
    private final IService server;
    private final Socket connection;
    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public ClientRpcReflectionWorker(IService server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = new ObjectOutputStream(connection.getOutputStream());
            output.flush();
            input = new ObjectInputStream(connection.getInputStream());
            connected = true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void run() {
        while (connected) {
            try {
                Object request = input.readObject();
                Response response = handleRequest((Request) request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException | ClassNotFoundException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error " + e);
        }
    }

    @Override
    public void ticketBought(Ticket ticket) throws Exception {
        try {
            sendResponse(new Response.Builder().type(ResponseType.TICKET_BOUGHT).data(ticket).build());
        } catch (IOException e) {
            throw new Exception("sending error: " + e);
        }
    }

    private static final Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    //  private static Response errorResponse=new Response.Builder().type(ResponseType.ERROR).build();
    private Response handleRequest(Request request) {
        Response response = null;
        String handlerName = "handle" + (request).type();
        System.out.println("HandlerName " + handlerName);
        try {
            Method method = this.getClass().getDeclaredMethod(handlerName, Request.class);
            response = (Response) method.invoke(this, request);
            System.out.println("Method " + handlerName + " invoked");
        } catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
            e.printStackTrace();
        }

        return response;
    }

    private Response handleLOGIN(Request request) {
        System.out.println("Login request ..." + request.type());
        System.out.println("Received request: " + request.data().toString());
        Employee employee = (Employee) request.data();
        try {
            Employee found = server.login(employee, this);
            return new Response.Builder().type(ResponseType.OK).data(found).build();
        } catch (Exception e) {
            connected = false;
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleLOGOUT(Request request) {
        System.out.println("Logout request...");
        Employee employee = (Employee) request.data();
        try {
            server.logout(employee, this);
            connected = false;
            return okResponse;

        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_A_FLIGHTS(Request request) {
        System.out.println("Get all flights request...");
        try {
            Collection<Flight> flights = server.getAllAvailableFlights();
            return new Response.Builder().type(ResponseType.GET_A_FLIGHTS).data(flights).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_FLIGHT(Request request) {
        System.out.println("Get flight request...");
        try {
            Flight flight = server.findFlightById((int) request.data());
            return new Response.Builder().type(ResponseType.GET_FLIGHT).data(flight).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleUPDATE_FLIGHT(Request request) {
        System.out.println("Update flight request...");
        try {
            Flight flight = (Flight) request.data();
            server.updateFlight(flight, flight.getId());
            return okResponse;
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_AIRPORT(Request request) {
        System.out.println("Get airport request...");
        try {
            Airport airport = server.findAirportById((int) request.data());
            return new Response.Builder().type(ResponseType.GET_AIRPORT).data(airport).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleGET_DD_FLIGHT(Request request) {
        System.out.println("Get DD flight request...");
        try {
            DTODestinationDate dds = (DTODestinationDate) request.data();
            Collection<Flight> flights = server.findFlightByDestinationDate(dds.getDestination(), dds.getDate());
            return new Response.Builder().type(ResponseType.GET_DD_FLIGHT).data(flights).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private Response handleBUY_TICKET(Request request){
        System.out.println("Buy ticket request...");
        try {
            DTOFlight dto = (DTOFlight) request.data();
            int numberOfSeats = server.buyTicket(dto.getClient(), dto.getPeople(), dto.getFlight());
            return new Response.Builder().type(ResponseType.OK).data(numberOfSeats).build();
        } catch (Exception e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }

    private void sendResponse(Response response) throws IOException {
        System.out.println("sending response " + response);
        output.writeObject(response);
        output.flush();
    }
}

File: Semester 4/Programming and Projection Enironments/mpp-proiect-csharp/Proiect_MPP/network/ClientRpcReflectionWorker.cs
using services;
using model;
using model.people;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using System.Threading.Tasks;
using System.Reflection;
using System.Collections.ObjectModel;

namespace network
{
    public class ClientRpcReflectionWorker : IObserver
    {
        private IService server;
        private TcpClient connection;

        private readonly NetworkStream stream;
        private readonly IFormatter formatter;
        private volatile bool connected;

        public ClientRpcReflectionWorker(IService server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
                formatter = new BinaryFormatter();
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

        public void ticketBought(List<Flight> flights)
        {
            try
            {
                sendResponse(new Response.Builder().Type(ResponseType.TICKET_BOUGHT).Data(flights).Build());
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
        }

        public virtual void Run()
        {
            while (connected)
            {
                try
                {
                    object request = formatter.Deserialize(stream);
                    object response = handleRequest((Request)request);
                    if (response != null)
                    {
                        sendResponse((Response)response);
                    }
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }

                try
                {
                    Thread.Sleep(1000);
                }
                catch (Exception e)
                {
                    Console.WriteLine(e.StackTrace);
                }
            }
            try
            {
                stream.Close();
                connection.Close();
            }
            catch (Exception e)
            {
                Console.WriteLine("Error " + e);
            }
        }

        private void sendResponse(Response response)
        {
            Console.WriteLine("sending response " + response);
            lock (stream)
            {
                formatter.Serialize(stream, response);
                stream.Flush();
            }
        }

        private Response handleRequest(Request request)
        {
            Response response = null;
            string handlerName = "handle" + request.Type;
            Console.WriteLine($"Handler name: {handlerName}");
            try
            {
                var method = GetType().GetMethod(handlerName, new[] { typeof(Request) });
                response = (Response)method.Invoke(this, new[] { request });
                Console.WriteLine("Method " + handlerName + " invoked");
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.StackTrace);
            }
            return response;
        }

        private static readonly Response okResponse = new Response.Builder().Type(ResponseType.OK).Build();

        public Response handleLOGIN(Request request)
        {
            Console.WriteLine("Login request ..." + request.Type);
            Console.WriteLine("Received request: " + request.Data.ToString());
            Employee employee = (Employee)request.Data;
            try
            {
                Employee found = server.login(employee, this);
                return new Response.Builder().Type(ResponseType.OK).Data(found).Build();
            }
            catch (Exception ex)
            {
                connected = false;
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.StackTrace).Build();
            }
        }

        public Response handleGET_FLIGHT(Request request)
        {
            Console.WriteLine("Get flight request...");
            try
            {
                Flight flight = server.findFlightById((int)request.Data);
                return new Response.Builder().Type(ResponseType.GET_FLIGHT).Data(flight).Build();
            }
            catch (Exception e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
            }
        }

        public Response handleLOGOUT(Request request)
        {
            Console.WriteLine("Logout request...");
            Employee employee = (Employee)request.Data;
            try
            {
                server.logout(employee, this);
                connected = false;
                return okResponse;
            }
            catch (Exception ex)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.StackTrace).Build();
            }
        }

        public Response handleGET_A_FLIGHTS(Request request)
        {
            Console.WriteLine("Get all flights request...");
            try
            {
                List<Flight> flights = server.getAllAvailableFlights().ToList();
                return new Response.Builder().Type(ResponseType.GET_A_FLIGHTS).Data(flights).Build();
            }
            catch (Exception e)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(e.StackTrace).Build();
            }
        }

        public Response handleGET_ALL_AIRPORTS(Request request)
        {
            Console.WriteLine("Get airports request...");
            try
            {
                List<Airport> airports = server.getAllAirports();
                return new Response.Builder().Type(ResponseType.GET_ALL_AIRPORTS).Data(airports).Build();
            }
            catch (Exception ex)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.StackTrace).Build();
            }
        }

        public Response handleGET_DD_FLIGHT(Request request)
        {
            Console.WriteLine("Get DD flight request...");
            try
            {
                DTODestinationDate dds = (DTODestinationDate)request.Data;
                List<Flight> flights = server.findFlightByDestinationDate(dds.Destination, dds.Date);
                return new Response.Builder().Type(ResponseType.GET_DD_FLIGHT).Data(flights).Build();
            }
            catch (Exception ex)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.StackTrace).Build();
            }
        }

        public Response handleBUY_TICKET(Request request)
        {
            Console.WriteLine("Buy ticket request...");
            try
            {
                DTOFlight dto = (DTOFlight)request.Data;
                server.buyTicket(dto.client, dto.people, dto.flight);
                return okResponse;
            }
            catch (Exception ex)
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data(ex.StackTrace).Build();
            }
        }
    }
}
